cores/esp8266/umm_malloc/umm_malloc.cpp

```diff
@@ -314,15 +431,90 @@ void umm_init( void ) {
     UMM_PBLOCK(UMM_BLOCK_LAST) = 1;
 }
 
+
+void umm_init_common( size_t id, void *start_addr, size_t size, bool zero ) {
+  /* Preserve internal setup */
+  umm_heap_context_t *_context = umm_get_heap_by_id(id);
+  if (NULL == start_addr || NULL == _context || _context->heap) {
+    return;
+  }
+
+  /* init heap pointer and size, and memset it to 0 */
+  _context->id        = id;
+  _context->heap      = (umm_block *)start_addr;
+  _context->heap_end  = (void *)((uintptr_t)start_addr + size);
+  _context->numblocks = (size / sizeof(umm_block));
+
+  // An option for blocking the zeroing of extra heaps allows for performing
+  // post-crash discovery.
+  if (zero) {
+  	memset(_context->heap, 0x00, size);
+#if (!defined(UMM_INLINE_METRICS) && defined(UMM_STATS)) || defined(UMM_STATS_FULL)
+    memset(&_context->stats, 0x00, sizeof(_context->stats));
+#endif
+
+    /* Set up internal data structures */
+    umm_init_stage_2(_context);
+  }
+}
+
+void umm_init( void ) {
+  // if (umm_heap) {
+  //   return;
+  // }
+  for (size_t i = 0; i < UMM_NUM_HEAPS; i++) {
+    heap_context[i].heap = NULL;
+  }
+  memset(&heap_context[0], 0, sizeof(heap_context));
+  umm_init_common( UMM_HEAP_DRAM, (void *)UMM_MALLOC_CFG_HEAP_ADDR, UMM_MALLOC_CFG_HEAP_SIZE, true );
+  // umm_heap = (void *)&heap_context;
+}
+
+#ifdef UMM_HEAP_IRAM
+#include <core_esp8266_non32xfer.h>
```

Comment:
This creates a circular dependency: the core depends on umm, and umm depends on the core. This is never a good thing.
Instead, I suggest one of the following:
1. a weak function that in here is implemented empty, and is implemented as calling install_non32xfer_exception_handler(). I'm not fond of adding additional weak symbols.
2. receive a function pointer as arg, to be called on init. I'm not sure this is possible.
3. use the umm config header to define a binding to an external init function to be called. The external function is implemented in the core.

Reply:
The need is to have the  non32xfer_exception_handler installed before creating an IRAM heap. 

There is already a call to install the non32xfer_exception_handler in `core_esp8266_main.cpp`. I can extend the #if around that call to include a test for `UMM_HEAP_IRAM`. 

This would mean no earlier initialization of IRAM heap before `user_init` runs. Which seems highly unlikely anyways. We already have the ability to handle an early init of the DRAM heap. Which is what the SDK needs at startup.